### ISSUE TITLE
Enable High DPI rendering on macOS (rendering is blurry)

### ISSUE THREAD

**Issue by TomSchimansky** [pygame/pygame#2853]
I know this has been discussed a few times but maybe there's a solution I didn't found or someone wants to work on this now.

So I used the tkinter canvas before, which renders completely fine on macOS with the full resolution and antialiasing. And now I wanted to switch to pygame but I noticed that it uses only half the resolution and therefore looks blurry on my 2880x1800 Retina screen. I found out that on Windows you can set the `SetProcessDpiAwareness` command to enable high DPI rendering but on macOS there is no equivalent, you can only set the `NSHighResolutionCapable` in the .plist of an .app but therefore you have to compile it to an .app bundle which is not my plan at the moment.

So why is it not possible with pygame on macOS to render with full quality? I can not imagine that the SDL2 framework is not capable of doing it? Because at the moment this is not usable for me, when you use a lot of images, the low quality isn't noticeable that much, but with thin lines it looks very bad I think.

I recreated the `aacircle.py` from the examples folder to compare the quality:
<img width="1201" alt="pygame render quality macOS 2" src="https://user-images.githubusercontent.com/66446067/142504903-2eaca489-503e-4eb2-a6b1-5595785e18c7.png">
Zoomed in:
<img width="590" alt="pygame render quality macOS" src="https://user-images.githubusercontent.com/66446067/142504930-2e068263-cb0b-49a6-9df9-43a07e0084c2.png">

Besides this, the performance of pygame on macOS is really good as I can tell from running a few example games. So I would love to use it for my projects.

**Comment by Starbuck5** [comment]
Hey again,

I don't think this is possible in pygame right now, but it might be easy to add.

Right now, high dpi functionality is only supported in the experimental pygame._sdl2.Window class, which you wouldn't be able to use for this use case yet either. (I'm working on improving the Window class so you could)

The SDL docs are a bit unclear on this: https://wiki.libsdl.org/SDL_WindowFlags

> On Apple's OS X you must set the NSHighResolutionCapable Info.plist property to YES, otherwise you will not receive a High DPI OpenGL canvas.

Does that only apply to windows with the SDL_WINDOW_OPENGL flag, or all windows?

Also keep in mind that quadrupling (or whatever) the number of pixels will have a significant performance impact.

**Comment by TomSchimansky** [comment]
Yes of course there are four times more pixels to render, but I'm confident this will be no problem, because I tested a game I made years ago in fullscreen mode with full resolution of 2880x1800 and it runs smoothly with 60 fps. In fullscreen mode it is already possible to increase the resolution to the maximum, but the mouse becomes really small, the scaling is wrong.

Maybe I can try to test if the SDL_WINDOW_OPENGL flag is enough for SDL to render in high DPI mode.

**Comment by Starbuck5** [comment]
>  if the SDL_WINDOW_OPENGL flag is enough for SDL to render in high DPI mode.

I interpreted the documentation as saying you definitely **cannot** get a high dpi display programmatically with OpenGL. And OpenGL in this case is not a backend, it's a separate way of rendering, rather than using a display surface you can do everything through OpenGL with the window.

Also cool to see you joined the discord.

**Comment by robertpfeiffer** [comment]
Should we attempt to add this to pygame.init() or to set_mode? I think the main problem is finding an API that works with Windows, Mac, and Wayland. If there was a consistent way, SDL2 itself would already have done it. 

> I interpreted the documentation as saying you definitely cannot get a high dpi display programmatically with OpenGL. 
OpenGL should *just work* with high DPI, right? Unless you render to texture, there is nothing to be done, OpenGL is already based on floating point coordinated between -1 and +1. What's the problem?

**Comment by williamhCode** [comment]
Currently I am working on a OpenGL project on a retina mac and there's virtually no way to get OpenGL to render at Native Resolution using pygame. I've tried resizing the display and looking through gl attributes. It would be nice if the SDL_WINDOW_ALLOW_HIGHDPI flag was added, probably best in the pygame.display.set_mode() flags.

**Comment by williamhCode** [comment]
> Yes of course there are four times more pixels to render, but I'm confident this will be no problem, because I tested a game I made years ago in fullscreen mode with full resolution of 2880x1800 and it runs smoothly with 60 fps. In fullscreen mode it is already possible to increase the resolution to the maximum, but the mouse becomes really small, the scaling is wrong.
> 
> Maybe I can try to test if the ~SDL_WINDOW_OPENGL~ SDL_WINDOW_ALLOW_HIGHDPI flag is enough for SDL to render in high DPI mode.

Any updates on this?

**Comment by robertpfeiffer** [comment]
@williamhCode The behaviour differs a bit between windows, wayland, and mac OS and also interacts with screen zoom, and people have strong opinions on this. As it works now, SCALED will make it so screen zoom on Windows doesn't do anything, and that's fine if you're blowing up GameBoy resolution to the whole screen, but some extra smartypants developer might use SCALED with a resolution of 1920x1080 to make sure his game is not enlarged by the accessibility tools that come with windows. High DPI and per-display zoom mostly doesn't work consistently on X11, but it should work on Wayland. I have tested it with sway.

**Comment by williamhCode** [comment]
@robertpfeiffer Hmm I see. I think for now, I will focus on getting Pygame with OpenGL to render at HighDPI. High DPI will always be a hot issue since behavior always differ from system to system. Nonetheless, I have used Pyglet and the OpenGL window works with High DPI, so getting it to work with Pygame is definitely possible. I would much prefer using Pygame over Pyglet since I like its structural style over Pyglet's OO style, so I am mainly working on this for my own use.

If anyone else wants to work on this then please join me. Also can someone rename the issue?

**Comment by williamhCode** [comment]
> Rename to what?

Probably something more accurate like "Enable High DPI rending on OpenGL window", since it works with normal windows.

**Comment by fladd** [comment]
Is there an update to this?
And how is the situation under Linux? Does it also only return a low-resolution (i.e. scaled) OpengGL display?

**Comment by segment-tree** [comment]
On my MacBook when I run pygame in linux(asahilinux) ,pygame think my resolution is 2k , but in macOS pygame think my resolution is 1k and the rendering is blurry as well

**Comment by SmoolPerson** [comment]
updates? please I am trying to get my mandelbrot rendering to work and it looks very pixelated :(

**Comment by taiyo66666-hash** [comment]
I managed to get sharp, high-res rendering working by creating a custom extension that overlays a Metal view on top of the Pygame window.

It's more of a bypass than a fix for the core SDL issue, but it works great for getting crisp 1:1 rendering on HiDPI screens. I've packaged it with a Pygame-like drawing API.

You can check it out here: [Pygame-like drawing API](https://github.com/taiyo66666-hash/MetalPygameDistribution)

**Comment by SmoolPerson** [comment]
Thank you so much you are a livesaver <3
